epix/simulator/helpers.py

```diff
@@ -0,0 +1,116 @@
+        SeparationDistanceSlope = \
+          5.5869678412887262e-07 * max_s2_area + 0.0044792968
+    SeparationDistance = \
+        SeparationDistanceIntercept - \
+        SeparationDistanceSlope * (-sensitive_volume_ztop +\
+            (z1 + z2) * 0.5)
+    return z1-z2<SeparationDistance
+
+class Helpers():
+    @staticmethod
+    def assignOrder(order):
+        """This is a trick to have the calculation function be executed in a particular order"""
+        def do_assignment(to_func):
+            to_func.order = order
+            return to_func
+
+        return do_assignment
+
+    @staticmethod
+    def average_spe_distribution(spe_shapes):
+        """We take the spe distribution from all channels and take the average to be our spe distribution to draw photon areas from"""
+        uniform_to_pe_arr = []
+        for ch in spe_shapes.columns[1:]:  # skip the first element which is the 'charge' header
+            if spe_shapes[ch].sum() > 0:
+                # mean_spe = (spe_shapes['charge'].values * spe_shapes[ch]).sum() / spe_shapes[ch].sum()
+                scaled_bins = spe_shapes['charge'].values  # / mean_spe
+                cdf = np.cumsum(spe_shapes[ch]) / np.sum(spe_shapes[ch])
+            else:
+                # if sum is 0, just make some dummy axes to pass to interpolator
+                cdf = np.linspace(0, 1, 10)
+                scaled_bins = np.zeros_like(cdf)
+
+            grid_cdf = np.linspace(0, 1, 2001)
+            grid_scale = interp1d(cdf, scaled_bins,
+                                  bounds_error=False,
+                                  fill_value=(scaled_bins[0], scaled_bins[-1]))(grid_cdf)
+
+            uniform_to_pe_arr.append(grid_scale)
+        spe_distribution = np.mean(uniform_to_pe_arr, axis=0)
+        return spe_distribution
+    
+    @staticmethod
+    @numba.njit
+    def macro_cluster_events(instructions):
+        """Loops over all instructions, checks if it's an s2 and if there is another s2 within the same event
+            within the macro cluster distance, if it is they are merged."""
+        for ix1,_ in enumerate(instructions):
+            if instructions[ix1]['type']!=2:
+                continue
+            for ix2 in range(1,len(instructions[ix1:])+1):
+                if instructions[ix1+ix2]['type']!=2:
+                    continue
+                if instructions[ix1]['event_number']!=instructions[ix1+ix2]['event_number']:
+                    break
+                if _merge_these_clusters(instructions[ix1]['amp'],instructions[ix1]['z'],
+                                         instructions[ix1+ix2]['amp'],instructions[ix1+ix2]['z']):
+                    instructions[ix1+ix2]['x'] = (instructions[ix1]['x']+instructions[ix1+ix2]['x'])*0.5
+                    instructions[ix1+ix2]['y'] = (instructions[ix1]['y']+instructions[ix1+ix2]['y'])*0.5
+                    instructions[ix1+ix2]['z'] = (instructions[ix1]['z']+instructions[ix1+ix2]['z'])*0.5
```

Comment:
I think it makes more sense to calculate the weighted average than just the mean.